File: Namyata/CheckIfEmailAddressIsUnique.py
# Every valid email consists of a local name and a domain name, separated by the '@' sign. Besides lowercase letters, the email may contain one or more '.' or '+'.
#
# For example, in "alice@leetcode.com", "alice" is the local name, and "leetcode.com" is the domain name.
# If you add periods '.' between some characters in the local name part of an email address, mail sent there will be forwarded to the same address without dots in the local name. Note that this rule does not apply to domain names.
#
# For example, "alice.z@leetcode.com" and "alicez@leetcode.com" forward to the same email address.
# If you add a plus '+' in the local name, everything after the first plus sign will be ignored. This allows certain emails to be filtered. Note that this rule does not apply to domain names.
#
# For example, "m.y+name@email.com" will be forwarded to "my@email.com".
# It is possible to use both of these rules at the same time.
#
# Given an array of strings emails where we send one email to each emails[i], return the number of different addresses that actually receive mails.


import unittest
from typing import List

def numUniqueEmails(emails: List[str]) -> int:
    set_ = set()
    for i in emails:
        final_email = ""
        email = i.split("@")
        email[0] = email[0].replace(".", "")
        if "+" in email[0]:
            index = email[0].index("+")
            email[0] = email[0][:index]
        final_email += email[0] + "@" + email[1]
        set_.add(final_email)
    return len(set_)

class TestNumUniqueEmails(unittest.TestCase):
    def test_empty_list(self):
        emails = []
        expected = 0
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_single_email(self):
        emails = ["test.email@example.com"]
        expected = 1
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_duplicate_emails(self):
        emails = ["test.email@example.com", "test.email@example.com"]
        expected = 1
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_unique_emails(self):
        emails = ["test.email@example.com", "test.email2@example.com", "another.email@example.com"]
        expected = 3
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_emails_with_dots(self):
        emails = ["t.est.email@example.com", "testemail@example.com"]
        expected = 1
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_emails_with_plus(self):
        emails = ["test.email+spam@example.com", "test.email@example.com"]
        expected = 1
        self.assertEqual(numUniqueEmails(emails), expected)

    def test_emails_with_com_suffix(self):
        emails = ["test.email@example.com", "test.email@example.org"]
        expected = 2
        self.assertEqual(numUniqueEmails(emails), expected)

if __name__ == '__main__':
    unittest.main()

#All unit tests are correct and passed

File: CTests/check_if_email_address_is_unique.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <string.h>

#define MAX_EMAIL_LENGTH 100

int numUniqueEmails(char emails[][MAX_EMAIL_LENGTH], int emails_length) {
    int unique_count = 0;
    char unique_emails[emails_length][MAX_EMAIL_LENGTH];

    for (int i = 0; i < emails_length; i++) {
        char final_email[MAX_EMAIL_LENGTH];
        char* email = strtok(emails[i], "@");
        strcpy(final_email, email);
        char* domain = strtok(NULL, "@");

        // Remove dots from the local name
        char* p = strchr(final_email, '.');
        while (p != NULL) {
            memmove(p, p + 1, strlen(p));
            p = strchr(final_email, '.');
        }

        // Ignore everything after the first '+' sign
        p = strchr(final_email, '+');
        if (p != NULL) {
            *p = '\0';
        }

        strcat(final_email, "@");
        strcat(final_email, domain);

        bool is_duplicate = false;
        for (int j = 0; j < unique_count; j++) {
            if (strcmp(final_email, unique_emails[j]) == 0) {
                is_duplicate = true;
                break;
            }
        }

        if (!is_duplicate) {
            strcpy(unique_emails[unique_count], final_email);
            unique_count++;
        }
    }

    return unique_count;
}

void test_empty_list() {
    char emails[][MAX_EMAIL_LENGTH] = {};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 0;
    assert(result == expected);
    printf("Test case 'empty_list' passed\n");
}

void test_single_email() {
    char emails[][MAX_EMAIL_LENGTH] = {"test.email@example.com"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 1;
    assert(result == expected);
    printf("Test case 'single_email' passed\n");
}

void test_duplicate_emails() {
    char emails[][MAX_EMAIL_LENGTH] = {"test.email@example.com", "test.email@example.com"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 1;
    assert(result == expected);
    printf("Test case 'duplicate_emails' passed\n");
}

void test_unique_emails() {
    char emails[][MAX_EMAIL_LENGTH] = {"test.email@example.com", "test.email2@example.com", "another.email@example.com"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 3;
    assert(result == expected);
    printf("Test case 'unique_emails' passed\n");
}

void test_emails_with_dots() {
    char emails[][MAX_EMAIL_LENGTH] = {"t.est.email@example.com", "testemail@example.com"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 1;
    assert(result == expected);
    printf("Test case 'emails_with_dots' passed\n");
}

void test_emails_with_plus() {
    char emails[][MAX_EMAIL_LENGTH] = {"test.email+spam@example.com", "test.email@example.com"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 1;
    assert(result == expected);
    printf("Test case 'emails_with_plus' passed\n");
}

void test_emails_with_com_suffix() {
    char emails[][MAX_EMAIL_LENGTH] = {"test.email@example.com", "test.email@example.org"};
    int emails_length = sizeof(emails) / sizeof(emails[0]);
    int result = numUniqueEmails(emails, emails_length);
    int expected = 2;
    assert(result == expected);
    printf("Test case 'emails_with_com_suffix' passed\n");
}

int main() {
    test_empty_list();
    test_single_email();
    test_duplicate_emails();
    test_unique_emails();
    test_emails_with_dots();
    test_emails_with_plus();
    test_emails_with_com_suffix();

    return 0;
}
